tests/acceptance/test_update.py

```diff
@@ -139,24 +184,133 @@ def test_network_based_image_update(self, successful_image_update_mender):
+        loses power. This causes the stored U-Boot environment to become
+        corrupt. On the next boot, U-Boot detects this and reverts to its built
+        in environment instead.
+
+        But this is a problem: The default environment will boot from rootfs
+        partition A, which contains a broken update. And since U-Boot at this
+        point doesn't know that an update was in progress, it will not attempt
+        to boot from anywhere else (`upgrade_available=0`). Hence the device is
+        bricked.
+
+        This is what a redundant U-Boot environment is supposed to protect
+        against by always providing two copies of the stored environment, and
+        guaranteeing that at least one of them is always valid.
+
+        In a test we cannot pull the power from the device reliably, but it's
+        quite easy to simulate the situation by setting up the above scenario,
+        and then corrupting the environment manually with a file write.
+
+        """
+
+        if not env.host_string:
+            # This means we are not inside execute(). Recurse into it!
+            execute(self.test_redundant_uboot_env, successful_image_update_mender, bitbake_variables, corrupt_old_env)
+            return
+
+        (active, passive) = determine_active_passive_part(bitbake_variables)
+
+        if active != bitbake_variables["MENDER_ROOTFS_PART_B"]:
+            # We are not running the secondary partition. This is a requirement
+            # for this test to test the correct scenario. Do a full update, so
+            # that we end up on the right partition. Run the full update test to
+            # correct this. If running all the tests in order with a fresh
+            # build, the correct partition will usually be selected already.
+            self.test_network_based_image_update(successful_image_update_mender, bitbake_variables)
+
+            (active, passive) = determine_active_passive_part(bitbake_variables)
+            assert(active == bitbake_variables["MENDER_ROOTFS_PART_B"])
+
+        # Make a note of the checksums of each environment. We use this later to
+        # determine which one changed.
+        old_checksums = Helpers.get_env_checksums(bitbake_variables)
+
+        orig_env = run("fw_printenv")
+
+        # Make a dummy/broken update
+        subprocess.call("dd if=/dev/zero of=image.dat bs=1M count=0 seek=8", shell=True)
+        subprocess.call("artifacts write rootfs-image -t %s -n test-update -u image.dat -o image.mender" % image_type, shell=True)
+        put("image.mender", remote_path="/var/tmp/image.mender")
+        run("mender -rootfs /var/tmp/image.mender")
+
+        new_checksums = Helpers.get_env_checksums(bitbake_variables)
+
+        # Exactly one checksum should be different.
+        assert(old_checksums[0] == new_checksums[0] or old_checksums[1] == new_checksums[1])
+        assert(old_checksums[0] != new_checksums[0] or old_checksums[1] != new_checksums[1])
```

Comment:
Hmm, not really, the if statement below won't catch the case if both are different (it will enter the first branch in this case). Perhaps it's better if I remove the `else` below, because that one should never be hit after these asserts.